fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/Binder.java

```diff
@@ -162,6 +153,76 @@ private GroupExpression extractGroupExpression(Pattern pattern, Group group) {
         }
     }
 
+    /**
+     * Expression binding for MULTI_JOIN, which contains only JOIN/SCAN nodes
+     * <p>
+     * NOTE: Why not match using regular pattern ?
+     * 1. Regular matching cannot bind the tree but only the partial expression. But MULTI_JOIN needs to extract
+     * the entire tree
+     * 2. Regular matching is extremely slow in this case since it needs to recursive descent the tree, build the
+     * binding state and check the expression at the same time. But MULTI_JOIN could enumerate the GE without any check
+     */
+    class MultiJoinBinder {
+
+        public OptExpression match(GroupExpression ge) {
+            // 1. Check if the entire tree is MULTI_JOIN
+            // 2. Enumerate GE
+            if (ge == null || !isMultiJoin(ge)) {
+                return null;
+            }
+
+            return enumerate(ge);
+        }
+
+        private OptExpression enumerate(GroupExpression ge) {
+            if (ge == null) {
+                return null;
+            }
+            List<OptExpression> resultInputs = Lists.newArrayList();
+
+            int groupExpressionIndex = 0;
+            while (groupExpressionIndex < ge.getInputs().size()) {
+                trace();
+                Group group = ge.getInputs().get(groupExpressionIndex);
+                OptExpression opt = enumerate(extractGroupExpression(group));
+
+                if (opt == null) {
+                    return null;
+                } else {
+                    resultInputs.add(opt);
+                }
+
+                groupExpressionIndex++;
+            }
+
+            return new OptExpression(ge, resultInputs);
+        }
+
+        private GroupExpression extractGroupExpression(Group group) {
+            int valueIndex = groupExpressionIndex.get(groupTraceKey);
+            if (valueIndex >= group.getLogicalExpressions().size()) {
+                groupExpressionIndex.remove(groupTraceKey);
+                return null;
+            }
+            return group.getLogicalExpressions().get(valueIndex);
+        }
+
+        private boolean isMultiJoin(GroupExpression ge) {
+            return ge.getOp().getOpType() == OperatorType.LOGICAL_JOIN && isMultiJoinRecursive(ge);
+        }
+
+        private boolean isMultiJoinRecursive(GroupExpression ge) {
+            if (!isMultiJoinOp(ge.getOp().getOpType())) {
+                return false;
+            }
+            return ge.getInputs().stream().allMatch(x -> isMultiJoinRecursive(x.getFirstLogicalExpression()));
```

Comment:
it seems wrong, why only check the first one, it seems that we should check every logicalExpression in the Group to see if any one's root operator matches either Scan|CTEConsume|Join

Reply:
yes. 

In normal case the first operator type could indicates the operator type of the group.

A special scenario is *Nested MV`. Let's say the two-level aggregation, the first level of aggregation would be rewritten as `SCAN`, so the Rule would put a `SCAN` in the original `AGG` group, so that the following procedure could rewrite the `AGG->SCAN` to a `SCAN`.

To handle this case, we need to check the `firstLogicalExpression` first, then any `SCAN` node.